kyo-prelude/shared/src/main/scala/kyo2/Memo.scala

```diff
@@ -0,0 +1,32 @@
+package kyo2
+
+import kyo.Tag
+import kyo2.kernel.*
+
+opaque type Memo = Var[Memo.Cache]
+
+object Memo:
+
+    private[kyo2] case class Cache(map: Map[(Any, Any), Any])
+
+    private val empty = Cache(Map.empty)
+
+    def apply[A, B, S](f: A => B < S)(using Frame): A => B < (S & Memo) =
+        val token = new Object
```

Comment:
What's the object for? I'm surprised this creates the behavior shown in the tests. Is the hashcode stable for a given A?

Reply:
It's used to ensure each memoized function has a different key space in the cache, otherwise values from one function could be returned by another. The `hashCode` is the identity of the object itself created with each function, unrelated to `A`. I've done some refactoring to make it clearer.